retry/retry.go

```diff
@@ -0,0 +1,142 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package retry provides a framework for retrying actions.
+// It does not itself invoke the action to be retried, but
+// is intended to be used in a retry loop.
+//
+// The basic usage is as follows:
+//
+//	for a := someStrategy.Start(); a.Next(); {
+//		try()
+//	}
+//
+// See examples for details of suggested usage.
+package retry
+
+import (
+	"time"
+
+	"github.com/juju/utils/clock"
+)
+
+// Strategy is implemented by types that represent a retry strategy.
+type Strategy interface {
+	NewTimer(now time.Time) Timer
+}
+
+// Timer represents a source of timing events for a retry strategy.
+type Timer interface {
+	// NextSleep returns the length of time to sleep
+	// before the next retry. If no more attempts should
+	// be made it should return false, and the returned
+	// duration will be ignored.
+	//
+	// Note that NextSleep is called once after
+	// each iteration has completed, assuming the
+	// retry loop is continuing.
+	NextSleep(now time.Time) (time.Duration, bool)
```

Comment:
When would you ever call this with anything other than time.Now()?  Isn't this why we have a fake clock implementation?